pack3d/pack3d.go

```diff
@@ -0,0 +1,356 @@
+	VolumeWithSpacing float64
+}
+
+// Attempts to pack a list of objects, applying rotation, scaling, and co-packing if specified.
+func Pack(config *Config) (*PackingOutput, error) {
+
+	var (
+		singleStlSize  []fauxgl.Vector
+		scaleStl       []fauxgl.Matrix
+		mfgRotationStl []fauxgl.Matrix
+		done           func()
+		totalVolume    float64
+		iterations     int
+		srcStlNames    []string
+		transMaps      []TransMap
+	)
+
+	model := NewModel()
+	scale := 1.0
+	var scaleMatrix fauxgl.Matrix
+	var mfgRotationMatrix fauxgl.Matrix
+
+	spacing := config.Spacing / 2.0
+
+	// frameSize is the vertex in the first quadrant
+	frameSize := fauxgl.V(config.BuildVolume[0], config.BuildVolume[1], config.BuildVolume[2])
+	buildVolume := config.BuildVolume[0] * config.BuildVolume[1] * config.BuildVolume[2]
+	//fmt.Println(frameSize)
+
+	/* Loading stl models */
+
+	// Tech Debt: there's unnecessary repetition of code inside the if-statement below.
+
+	coPackMap := make(map[string][]*Copack) // variable that contains co-packed mesh's data.
+	for _, item := range config.ConfigItems {
+
+		var mesh *fauxgl.Mesh
+		var err error
+
+		if item.Copack == nil {
+
+			// 1. load the mesh.
+			done = timed(fmt.Sprintf("loading mesh %s", item.Filename))
+			mesh, err = fauxgl.LoadMesh(item.Filename)
+			if err != nil {
+				return nil, err
+			}
+			done()
+
+			// 2. mesh centering.
+			mesh.Center()
+
+			// 3. apply the scaling to the mesh.
+			//    Notice that if scaling is to be applied, it is done
+			//    before the computation of the BoundingBox and volume.
+			scale = item.Scale
+			scaleMatrix = fauxgl.Scale(fauxgl.V(scale, scale, scale))
+			if scale != 1.0 {
+				done = timed("scaling mesh")
+				mesh.Transform(scaleMatrix)
+				done()
+			}
+
+			// 4. apply the manufacturing rotation mesh.
+			//    Notice that this is done before the computation of the BoundingBox and volume.
+			// IMPORTANT: do not confuse manufacturing orientation with the packing
+			//            orientations from the orientations provided by the annealing further on.
+			mfgRotationMatrix = item.ManufacturingOrientation()
+			mesh.Transform(mfgRotationMatrix)
+
+			// 5. update all the copies mesh for the json output.
+			size := mesh.BoundingBox().Size()
+			for i := 0; i < item.Count; i++ {
+				singleStlSize = append(singleStlSize, size)
+				srcStlNames = append(srcStlNames, item.Filename)
+				scaleStl = append(scaleStl, scaleMatrix)
+				mfgRotationStl = append(mfgRotationStl, mfgRotationMatrix)
+			}
+
+			fmt.Printf("  %d triangles\n", len(mesh.Triangles))
+			fmt.Printf("  %g x %g x %g\n", size.X, size.Y, size.Z)
+
+			// 6. coarse approx of its volume.
+			totalVolume += mesh.BoundingBox().Volume()
+
+		} else {
+
+			coPackMap[item.Filename] = item.Copack
+
+			// 1a. load the main co-packing mesh (the "parent" co-packing mesh, so to say).
+			done = timed(fmt.Sprintf("loading the main co-packing mesh %s", item.Filename))
+			mesh, err = fauxgl.LoadMesh(item.Filename)
```

Comment:
Have only just realised that Fogelman called this library as a pun on their name...

Reply:
...same here...